### PR TITLE
www: Implement copy button for rustup download command

### LINKED ISSUE #2280: Copy button for Unix installation command on Rustup.rs
**Describe the problem you are trying to solve**

On the main [rustup.rs](https://rustup.rs/) installation site, the Unix installation command is very long and copying it requires having to click and drag while scrolling or triple-clicking.

unix command

**Describe the solution you'd like**

Having a small button or intractable area that copies the command to the clipboard (and it provides feedback that it has done so) would make the installation process easier.

**Examples**

[Crates.io](https://crates.io/crates/lru) crate pages:
lru

[Npm](https://www.npmjs.com/package/react) package pages:
react

**Notes**

The fact that the crates.io team has already implemented this makes it seem like using/looking at their implementation would be worthwhile.

### PR DESCRIPTION
**Pull request by BeniCheni**
Implement a "_Copy_" button in www/ site page (inspired by [crates.io](https://crates.io/), to fix #2280 

Screen Shot 2020-05-10 at 5 50 53 PM

Screen Shot 2020-05-10 at 5 51 02 PM

### CODE REVIEW

**Review by kinnison — COMMENTED** [review]
I think this looks okay, but I'm not an HTML/CSS/JS expert.  I wonder if @pietroalbini might know if this is sane?

**Comment by emilyalbini** [comment]
I'm not familiar with clipboard handling unfortunately :(

**Comment by kinnison** [comment]
Sadly the crates.io impl looks to be Ember based so I have no clue how to compare it.

**Comment by kinnison** [comment]
I've just tested this locally and I get: TypeError: 'clipboard-write' (value of 'name' member of PermissionDescriptor) is not a valid value for enumeration PermissionName.

Was this tested in Firefox at all @BeniCheni ?

**Comment by workingjubilee** [comment]
This does not need to use the Permissions API on Firefox, however Firefox requires a secure context instead. Please consult https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API for further details.

EDIT: Wrong link, whoops.

File: www/rustup.js
Comment (workingjubilee) [inline www/rustup.js]: Providing a second function to Promise.then() will catch Promise ~~errors~~ rejections but this does not catch logical errors in the JavaScript code on browsers which do not implement this API. This needs to be wrapped in a try / catch block.

File: www/rustup.js
Comment (workingjubilee) [inline www/rustup.js]: For a second, I thought we were somehow deleting the copied text itself (i.e. the text we just wrote to the clipboard). It would be nice if this function used a name which indicated clearly its usage of "this function is resetting the message that the user has copied the text." I'd also have it linger slightly longer.

File: www/index.html
Comment (workingjubilee) [inline www/index.html]: This should have an id more like "copy-status-message" or something, I think, see also the other comment re: the later functions involving it ("text" is ambiguous).

**Review by workingjubilee — CHANGES_REQUESTED** [review]

**Comment by BeniCheni** [comment]
> Was this tested in Firefox at all @BeniCheni ?

>  Please consult https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API for further details.

@kinnison / @workingjubilee, updated to use the `Clipboard.writeText()`, and tested in FireFox / Safari / Chrome. Worked fine in my local.

> This needs to be wrapped in a try / catch block.

> It would be nice if this function used a name which indicated clearly its usage of "this function is resetting the message that the user has copied the text." I'd also have it linger slightly longer.

> This should have an id more like "copy-status-message" or something,

Thanks for the feedback, @workingjubilee. Updated to use `try` / `catch`, renamed to `clear_copy_status_message` function w. 5 second timeout to clear copy status message, and used "copy-status-message" as element ID for the copy button.

PTAL? 🙏

**Review by workingjubilee — APPROVED** [review]
Excellent! I have reviewed this on my local and while (due to the vagaries of systems) I recommend this get yet another set of eyes on it before it is merged (hi Kinnison) it looks good and workable. At some point the unstable Web APIs involved will settle and this will need to be redrafted but there is no telling  whether that happens ~3 days, ~3 years, or ~3 decades from now.

**Comment by kinnison** [comment]
It now works for me on Firefox on Linux so let's assume it'll mostly work anywhere.  Thanks @BeniCheni and thanks @workingjubilee for helping and reviewing.

**Pull request merged**
